x/p2p/client.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	ErrAppRequestFailed = errors.New("app request failed")
+	ErrRequestPending   = errors.New("request pending")
+	ErrNoPeers          = errors.New("no peers")
+)
+
+// AppResponseCallback is called upon receiving an AppResponse for an AppRequest
+// issued by Client.
+// Callers should check [err] to see whether the AppRequest failed or not.
+type AppResponseCallback func(
+	nodeID ids.NodeID,
+	responseBytes []byte,
+	err error,
+)
+
+// CrossChainAppResponseCallback is called upon receiving an
+// CrossChainAppResponse for a CrossChainAppRequest issued by Client.
+// Callers should check [err] to see whether the AppRequest failed or not.
+type CrossChainAppResponseCallback func(
+	chainID ids.ID,
+	responseBytes []byte,
+	err error,
+)
+
+type Client struct {
+	handlerID uint8
+	router    *Router
+	sender    common.AppSender
+}
+
+// AppRequestAny issues an AppRequest to an arbitrary node decided by Client.
+// If a specific node needs to be requested, use AppRequest instead.
+// See AppRequest for more docs.
+func (c *Client) AppRequestAny(
+	ctx context.Context,
+	appRequestBytes []byte,
+	onResponse AppResponseCallback,
```

Comment:
In a sense we are adding middleware layer for the client with these callbacks. This could be a followup but would it make sense to track breaking that functionality out? For example it might make sense for different observabilty reasons to chain the callbacks and or adding `onRequest` (client metrics).

Reply:
I think I would like to do this as a follow-up, but I do think client metrics are definitely worth thinking about as you mentioned. I -think- we can do this with the current code by either providing a wrapped appsender to the router, or by wrapping this client, but it's something I'd need to think about. I agree there is definitely value in having metrics in the client.